mikrotik/resource_dns_record.go

```diff
@@ -36,64 +36,50 @@ func resourceRecord() *schema.Resource {
 }
 
 func resourceServerCreate(d *schema.ResourceData, m interface{}) error {
-	address := d.Get("address").(string)
-	name := d.Get("name").(string)
-	ttl := d.Get("ttl").(int)
+	record := prepareDnsRecord(d)
 
 	c := m.(client.Mikrotik)
 
-	record, err := c.AddDnsRecord(name, address, ttl)
+	dnsRecord, err := c.AddDnsRecord(record)
 	if err != nil {
 		return err
 	}
 
-	recordToData(record, d)
-	return nil
+	return recordToData(dnsRecord, d)
 }
 
 func resourceServerRead(d *schema.ResourceData, m interface{}) error {
 	c := m.(client.Mikrotik)
 
 	record, err := c.FindDnsRecord(d.Id())
 
-	// TODO: Ignoring this error can cause all resources to think they
-	// need to be created. We should more appropriately handle this. The
-	// error where the DNS record is not found is not actually an error and
-	// needs to be disambiguated from real failures
-	if err != nil {
+	if _, ok := err.(*client.NotFound); ok {
 		d.SetId("")
 		return nil
 	}
 
-	recordToData(record, d)
-	return nil
+	return recordToData(record, d)
 }
 
 func resourceServerUpdate(d *schema.ResourceData, m interface{}) error {
 	c := m.(client.Mikrotik)
 
-	address := d.Get("address").(string)
-	ttl := d.Get("ttl").(int)
-	name := d.Id()
-
-	record, err := c.FindDnsRecord(name)
+	currentRecord, err := c.FindDnsRecord(d.Id())
+	record := prepareDnsRecord(d)
+	record.Id = currentRecord.Id
```

Comment:
So this cannot be resolved at this moment. The reason why this has to remain (for now) is because the `d.Id()` is set as the `name` field--without retrieving the dns record, the `.id` mikrotik field cannot be fetched and updates fail. I was planning to fix this behavior in a follow up PR, making all of our resources `Id` in state be the `.id` mikrotik field (as we had discussed). This will provide standardization in the code + remove these duplicate client calls.

Reply:
Please create a github issue before merging so we make sure we track the issue :)